front: migrate to ESLint flat config

emersion opened the pull request:
ESLint v9 requires flat configs, it no longer supports the legacy
config file format. Migrate our configuration file to prepare for
the bump.

--ext has been removed from the package.json script because ESLint
rejects that CLI option when used with flat configs: file extensions
are now specified in the config file directly.

Depends on https://github.com/OpenRailAssociation/osrd/pull/13105  
Part of https://github.com/OpenRailAssociation/osrd/issues/12395  
See also https://github.com/OpenRailAssociation/osrd/pull/13134

emersion committed:
front: migrate to ESLint flat config

ESLint v9 requires flat configs, it no longer supports the legacy
config file format. Migrate our configuration file to prepare for
the bump.

--ext has been removed from the package.json script because ESLint
rejects that CLI option when used with flat configs: file extensions
are now specified in the config file directly.

Signed-off-by: Simon Ser <contact@emersion.fr>

SharglutDev commented on front/eslint.config.js:
```diff
@@ -0,0 +1,181 @@
+      react: {
+        version: 'detect',
+      },
+      'import/resolver': {
```
No need for this part anymore ? 
```
"node": {
        "paths": ["src"],
        "moduleDirectory": ["node_modules", "src"]
      },
```

emersion replied:
I don't think this is required anymore, but this is technically unrelated to the config file migration, so let's keep `node` for now!

SharglutDev commented on front/eslint.config.js:
```diff
@@ -0,0 +1,181 @@
+        },
+      ],
+
+      '@typescript-eslint/no-explicit-any': 2,
```
We don't have to specify overrides anymore ? As long as these rules are defined after the default ones of the plugin they get overridden ?

emersion replied:
The overrides were used in the old config file to only apply the rule to `*.ts` files (and skip it for `*.js`), but that's no longer necessary.

There is still an override below for generated OpenAPI files.

SharglutDev reviewed (approved):
Great job, thanks for handling this painful migration !

Just 2 minor questions

SharglutDev commented:
Something went wrong in the rebase ? The commit name /description doesn't seem to be the good one :)

emersion commented:
> Something went wrong in the rebase ? The commit name /description doesn't seem to be the good one :)

Oops, fixed now!

Synar commented on front/eslint.config.js:
```diff
@@ -0,0 +1,185 @@
+    languageOptions: {
+      parserOptions: {
+        projectService: true,
+        tsconfigRootDir: import.meta.dirname,
```
I'm curious, was this part not needed for the old config format?

Synar replied:
Also, should we still have "ecmaVersion": "latest"?

emersion replied:
This is taken from https://typescript-eslint.io/getting-started/typed-linting

The recommended flat config has this field set, the recommended legacy config doesn't…

Indeed, [`ecmaVersion`](https://typescript-eslint.io/packages/parser/#ecmaversion) seems to default to `es2018`, so it sounds like a good idea to retain the previous `latest` value, done!

Synar reviewed (approved):
LGTM!

emersion commented:
Hmm, unfortunately `@rollup/eslint` doesn't support flat configs. See https://github.com/rollup/plugins/discussions/1607

emersion commented:
Ref https://github.com/rollup/plugins/pull/1898

emersion committed:
ui: upgrade @rollup/plugin-eslint to v9.1

This is required to get support for flat config files:
https://github.com/rollup/plugins/pull/1898

Signed-off-by: Simon Ser <contact@emersion.fr>

emersion commented:
The rollup plugin PR linked above has been merged, so this can progress now!
